File: packages/import/src/mto.js
/**
 * The Material Takeoff — what the material team buys from.
 *
 * An FMR asks the warehouse to fetch material somebody has already bought.
 * An MTO comes earlier and asks a different question: this is what the work
 * package needs, go and quote it. It is the first document of the project,
 * before any FMR exists.
 *
 *   drawings arrive → MTO → quoted and ordered → material arrives
 *                                                      ↓
 *                             FMRs published → crews fetch it
 *
 * The same drawing scan feeds both, so an MTO costs no extra reading. What it
 * adds is the pipe schedule (the buyer cannot order wall thickness without
 * it) and a split by how material is bought: bolts and gaskets come from
 * different suppliers on different lead times than pipe and fittings, which
 * is why the takeoff form has a sheet for each.
 *
 * Written as CSV rather than xlsx deliberately. The workbook template belongs
 * to the client and varies by project; a CSV opens in Excel, pastes into
 * whatever template is current, and needs no dependency — the xlsx package is
 * unmaintained and this system stays free of it.
 */

import { LedgerError } from '../../core/src/domain/ledger.js';

/**
 * The takeoff form's own columns, in its own order.
 *
 * Taken from the client's real filled-in workbook
 * (`Archive/templates/createMTO/`), not inferred. The three paint columns are
 * blank in every example seen so far, but they are on the form the purchasing
 * team reads, and a missing column shifts everything after it on paste.
 */
const COLUMNS = Object.freeze([
  ['cwa', 'CWA'],
  ['iwp', 'IWP'],
  ['lineNumber', 'LINE NUMBER'],
  ['sheet', 'SHEET'],
  ['pipeSpec', 'PIPE SPEC'],
  ['description', 'DESCRIPTION'],
  ['size', 'SIZE'],
  ['commodityCode', 'COMMODITY CODE'],
  ['quantity', 'QTY'],
  ['uom', 'UOM'],
  ['epicPaintCode', 'EPIC PAINT CODE'],
  ['custPaintCode', 'CUST. PAINT CODE'],
  ['color', 'COLOR']
]);

/**
 * The sheets of the form, in the order the client's workbook has them.
 *
 * COMBINED holds every row; the category sheets are the same material sorted
 * by how it is bought. Blinds, valves and birdscreens are quoted separately
 * from pipe, which is why they get sheets of their own rather than being
 * lumped in with fittings.
 */
export const TAKEOFF_SHEETS = Object.freeze([
  'COMBINED',
  'BLINDS',
  'PIPE & FITTINGS',
  'BOLTS & GASKETS',
  'SUPPORTS',
  'VALVES',
  'BIRDSCREENS',
  'OTHER MATERIALS'
]);

const BLIND_RE = /\bBLINDS?\b/i;
const BIRDSCREEN_RE = /\bBIRD\s*SCREENS?\b/i;
const VALVE_RE =
  /(?:\bVALVES?\b|^\s*(?:BALL|CHECK|GATE|GLOBE|BUTTERFLY|DIAPHRAGM|PLUG|NEEDLE|CONTROL|RELIEF|SAFETY)\b)/i;
const PIPE_FITTING_RE =
  /\b(?:PIPE|ELL|ELBOW|TEE|REDUCER|CAP|COUPLING|SOCKOLET|WELDOLET|THREDOLET|NIPPLE|FLANGE|UNION|SWAGE)\b/i;

/**
 * Which sheet a row belongs on.
 *
 * Ported from the client's own `material_category_sheet`, including the order
 * of the tests — a blind is a blind before it is anything else, and a ball
 * valve is a valve rather than a fitting. Changing the order silently moves
 * material onto the wrong buyer's sheet.
 */
export function takeoffSheetFor(itemType, description) {
  const type = String(itemType ?? '').trim().toUpperCase();
  const text = String(description ?? '').replace(/\s+/g, ' ').trim();

  if (BLIND_RE.test(text)) return 'BLINDS';
  if (['BOLT', 'GASKET', 'WASHER'].includes(type)) return 'BOLTS & GASKETS';
  if (type === 'SUPPORT') return 'SUPPORTS';
  if (BIRDSCREEN_RE.test(text)) return 'BIRDSCREENS';
  if (VALVE_RE.test(text)) return 'VALVES';
  if (['PIPE', 'FITTING'].includes(type) || PIPE_FITTING_RE.test(text)) {
    return 'PIPE & FITTINGS';
  }
  return 'OTHER MATERIALS';
}

/**
 * One CSV cell.
 *
 * A commodity code like `5UGSP-02-15` is text, but a bare `1-2` is read by
 * Excel as a date — the same class of damage that turns 3/4" into 4-Mar on the
 * way in. Quoting every non-empty cell costs nothing and stops it.
 */
function cell(value) {
  const text = value == null ? '' : String(value);
  if (text === '') return '';
  return `"${text.replace(/"/g, '""')}"`;
}

/** One sheet of the takeoff, as CSV. */
export function takeoffCsv(rows) {
  const lines = [COLUMNS.map(([, label]) => cell(label)).join(',')];
  for (const row of rows) {
    lines.push(COLUMNS.map(([key]) => cell(row[key])).join(','));
  }
  // CRLF: Excel is the destination, and it is what the format specifies.
  return lines.join('\r\n');
}

/**
 * Group takeoff rows the way the form does.
 *
 * Every sheet is present even when empty — a buyer opening the file needs to
 * see that there are no bolts on this package, not wonder whether the sheet
 * failed to generate.
 */
export function groupBySheet(rows) {
  const grouped = new Map(TAKEOFF_SHEETS.map((name) => [name, []]));

  for (const row of rows ?? []) {
    // COMBINED is the whole takeoff; the category sheets are the same rows
    // sorted by who quotes them. A row appears on both, as it does on the
    // client's own workbook.
    grouped.get('COMBINED').push(row);

    const sheet = takeoffSheetFor(row.itemType, row.description);
    grouped.get(grouped.has(sheet) ? sheet : 'OTHER MATERIALS').push(row);
  }

  return grouped;
}

/**
 * The whole takeoff as one CSV, sheets in sequence.
 *
 * One file rather than three: the buyer forwards a single attachment, and the
 * sheet headings survive a paste into the client's own template.
 */
export function takeoffDocument({ rows, iwpNumber, cwa, drawings, missingPipeSpec }) {
  if (!rows?.length) {
    throw new LedgerError(
      'Those drawings hold no material to take off.', 'NO_MATERIAL'
    );
  }

  const grouped = groupBySheet(rows);
  const parts = [];

  parts.push([
    cell('MATERIAL TAKEOFF'),
    cell(iwpNumber || ''),
    cell(cwa ? `CWA ${cwa}` : ''),
    cell(`${drawings} drawing${drawings === 1 ? '' : 's'}`),
    cell(`${rows.length} line${rows.length === 1 ? '' : 's'}`)
  ].join(','));

  // Pipe nobody can order is worth saying at the top, not discovering at the
  // quoting stage.
  if (missingPipeSpec) {
    parts.push(cell(
      `${missingPipeSpec} pipe line${missingPipeSpec === 1 ? '' : 's'} `
      + 'had no pipe schedule on the drawing — check before ordering.'
    ));
  }

  for (const name of TAKEOFF_SHEETS) {
    const sheetRows = grouped.get(name);
    parts.push('', cell(name), takeoffCsv(sheetRows));
    if (!sheetRows.length) parts.push(cell('(nothing on this package)'));
  }

  return parts.join('\r\n');
}

/** What to call the file, so a folder of them stays readable. */
export function takeoffFilename({ iwpNumber, cwa }) {
  const stem = [cwa, iwpNumber].filter(Boolean).join('-') || 'package';
  return `MTO ${stem.replace(/[^A-Za-z0-9._-]+/g, '_')}.csv`;
}

File: packages/extract-iso/src/iso_bom/mto.py
"""Material Takeoff: the form the material team buys from.

An FMR asks the warehouse for material that has already been bought. An MTO
comes earlier — it is what the planner sends to the material manager and the
purchasing team so the material can be quoted and ordered in the first place.
Same drawings, same BOM rows, different question: the FMR asks "fetch this",
the MTO asks "buy this".

Because it is the same scan, this reads the pages the FMR extractor has
already classified and adds the two things the takeoff form needs and the
requisition does not:

  * **pipe schedule**, printed on the drawing, which the buyer needs to order
    the right wall thickness;
  * **item type**, so bolts and gaskets can be quoted separately from pipe and
    fittings — they are bought from different suppliers on different lead
    times, which is why the form has a sheet each.

The classification rules are ported from the MTO tool that produced the
takeoff workbooks this project was built against. They are not tidy, and they
are not guesses: each pattern is there because a real drawing wrote a material
that way.
"""

from __future__ import annotations

import re
from typing import Dict, List, Optional, Sequence

from .pdf_parser import _display_words, _join, _line_groups

# A schedule is short and alphanumeric — "40", "80", "10S", "315", "XS".
PIPE_SCHEDULE_VALUE_RE = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._/\-]*$")

# The trailing "-04" of LP131-HHWR-1091320-04 is the sheet, not part of the
# line number. The MTO form keeps them in separate columns.
SHEET_RE = re.compile(r"-(\d{1,2})\s*$")

# Fittings a buyer orders from a fittings supplier.
STANDARD_FITTING_RE = re.compile(
    r"^\s*(?:"
    r"ELL|ELBOW|TEE|RED|REDUCER|CONC\s+RED|ECC\s+RED|CAP|"
    r"COUPLING|COUP|CPLG|NIPPLE|NIP|FLG|FLANGE|OLET|"
    r"WELDOLET|SOCKOLET|THREDOLET|UNION|VALVE"
    r")(?:\b|[-,])",
    re.IGNORECASE,
)

# Bolts and gaskets, which are quoted separately and arrive separately.
STANDALONE_BOLT_RE = re.compile(
    r"^\s*(?:STUD[ -]?BOLT|MACHINE[ -]?BOLT|ANCHOR[ -]?BOLT|"
    r"HEX[ -]?BOLT|CAP[ -]?SCREW)\b",
    re.IGNORECASE,
)
STANDALONE_GASKET_RE = re.compile(r"^\s*GASKETS?\b", re.IGNORECASE)

# Turner's support commodity codes all start 5-something. A support is counted
# hardware, never measured pipe, however much its description mentions pipe.
# Support and attachment codes. All of these are counted hardware whose
# description names the pipe they hold — "WELDED SHOE LONG, SS, 3" HIGH, 12"
# PIPE" is a shoe, not twelve inches of pipe. Reading one as pipe is the same
# fault that once sent 344 rows of hardware to be quoted by the foot, so the
# code is the authority and the description is never consulted.
#
# Every entry below appears in the client's own newFmr36 package.
SUPPORT_CODE_RE = re.compile(
    r"^5(?:UGSP|UG|US|MUGSP|MUG|FSS\d*|FS\d*|MHR\d*|ABS\d*|MS|CH|CC|CI|C"
    r"|SH|S\d|DA\d*|ISC|BG\d*|MG\d*|G\d*)",
    re.IGNORECASE,
)

# Actual pipe, which is bought by the foot rather than counted.
PIPE_RE = re.compile(r"^\s*PIPE\b", re.IGNORECASE)

# A pipe quantity carries a foot mark: 19.1'
PIPE_LENGTH_RE = re.compile(r"^\s*([\d.]+)\s*'?\s*$")

# Where a takeoff row goes on the form.
SHEET_BOLTS_GASKETS = "BOLTS & GASKETS"
SHEET_PIPE_FITTINGS = "PIPE & FITTINGS"
SHEET_COMBINED = "COMBINED"

CWA_STOP_TOKENS = {
    "DATE", "DISCIPLINE", "IWP", "NAME", "PURPOSE", "REV",
    "SCOPE", "SIGNATURE", "STAGE", "STATUS", "TITLE",
}


def _clean_label(text: str) -> str:
    return re.sub(r"[^A-Z]", "", (text or "").upper())


def pipe_schedule(page) -> str:
    """The schedule printed on the drawing, or "" when it is not there.

    The buyer cannot order pipe without it, so a blank is worth surfacing
    rather than filling in with a guess.
    """
    for line in _line_groups(_display_words(page), tolerance=5.0):
        tokens = [_clean_label(word.text) for word in line]
        for index in range(len(tokens) - 1):
            if tokens[index:index + 2] != ["PIPE", "SCHEDULE"]:
                continue
            for word in line[index + 2:]:
                value = word.text.strip().rstrip(":")
                if PIPE_SCHEDULE_VALUE_RE.match(value):
                    return value
            return ""
    return ""


def cwa(page) -> str:
    """The CWA (construction work area) from a package cover page.

    Only trusted when the cover names exactly one — a cover listing several is
    ambiguous, and the caller can supply the value instead.
    """
    candidates: List[str] = []
    for line in _line_groups(_display_words(page), tolerance=5.0):
        for index, word in enumerate(line):
            if _clean_label(word.text) != "CWA":
                continue
            rest = []
            for following in line[index + 1:]:
                if _clean_label(following.text) in CWA_STOP_TOKENS:
                    break
                if following.text.endswith(":"):
                    break
                rest.append(following)
            value = _join(rest).strip(": ").strip()
            if value:
                candidates.append(value)

    unique = sorted(set(candidates))
    return unique[0] if len(unique) == 1 else ""


def sheet_number(drawing_number: str) -> str:
    """The sheet, taken off the end of the drawing number."""
    match = SHEET_RE.search(drawing_number or "")
    return match.group(1) if match else ""


def line_number(drawing_number: str) -> str:
    """The drawing number without its sheet suffix."""
    return SHEET_RE.sub("", drawing_number or "").strip()


def item_type(description: str, commodity_code: str) -> str:
    """What kind of material this is, for the purpose of buying it.

    Order matters. A support whose description says "U-BOLT GUIDE FOR 2\" PIPE"
    names the pipe it holds, not pipe being bought — reading it as pipe once
    sent 344 rows of hardware to be quoted by the foot.
    """
    text = description or ""
    code = commodity_code or ""

    # Supports first, before anything reads their description as pipe.
    if SUPPORT_CODE_RE.match(code):
        return "SUPPORT"
    if STANDALONE_BOLT_RE.match(text):
        return "BOLT"
    if STANDALONE_GASKET_RE.match(text):
        return "GASKET"
    if PIPE_RE.match(text):
        return "PIPE"
    if STANDARD_FITTING_RE.match(text):
        return "FITTING"
    return "SPECIALTY"


def takeoff_sheet(kind: str) -> str:
    """Which sheet of the takeoff form a row belongs on."""
    if kind in ("BOLT", "GASKET"):
        return SHEET_BOLTS_GASKETS
    if kind in ("PIPE", "FITTING"):
        return SHEET_PIPE_FITTINGS
    return SHEET_COMBINED


def takeoff_row(material: Dict[str, object], drawing: Dict[str, object],
                package: Dict[str, object]) -> Dict[str, object]:
    """One row of the takeoff form, in the columns the form actually has."""
    kind = item_type(material.get("description"), material.get("commodityCode"))
    quantity = str(material.get("quantity") or "").strip()
    uom = "EA"

    # Pipe is bought by the foot, and the drawing writes it with a foot mark.
    if kind == "PIPE":
        uom = "LF"
        length = PIPE_LENGTH_RE.match(quantity)
        if length:
            quantity = length.group(1)

    drawing_number = str(drawing.get("drawingNumber") or "")
    return {
        "cwa": package.get("cwa") or "",
        "iwp": package.get("iwpNumber") or "",
        # The client's form carries the whole drawing number here, sheet
        # suffix included, and repeats the sheet in its own column. Stripping
        # the suffix produced a value their purchasing team does not recognise.
        "lineNumber": drawing_number or line_number(drawing_number),
        "sheet": sheet_number(drawing_number),
        "pipeSpec": drawing.get("pipeSchedule") or "",
        "description": material.get("description") or "",
        "size": material.get("nominalSize") or "",
        "commodityCode": material.get("commodityCode") or "",
        "quantity": quantity,
        "uom": uom,
        "itemType": kind,
        "takeoffSheet": takeoff_sheet(kind),
    }


def takeoff(payload: Dict[str, object], cwa_override: Optional[str] = None
            ) -> Dict[str, object]:
    """Every drawing in a scanned package, as takeoff rows.

    Takes the payload `fmr_json.scan` already produces, so reading a package
    for an MTO and reading it for FMRs is one pass over the same PDFs.
    """
    package = {
        "iwpNumber": payload.get("iwpNumber") or "",
        "cwa": (cwa_override or payload.get("cwa") or "").strip(),
    }

    rows: List[Dict[str, object]] = []
    for drawing in payload.get("drawings") or []:
        for material in drawing.get("materials") or []:
            rows.append(takeoff_row(material, drawing, package))

    counts: Dict[str, int] = {}
    for row in rows:
        counts[row["takeoffSheet"]] = counts.get(row["takeoffSheet"], 0) + 1

    return {
        "iwpNumber": package["iwpNumber"],
        "cwa": package["cwa"],
        "rows": rows,
        "counts": counts,
        "drawings": len(payload.get("drawings") or []),
        # A row the buyer cannot act on: no schedule to order pipe against.
        "missingPipeSpec": sum(
            1 for row in rows if row["itemType"] == "PIPE" and not row["pipeSpec"]
        ),
    }
